File: src/lib/convert.ts
import { hex } from './entity.js';
import { countsByGraphic, diffCounts, type Change, type Counts } from './pack.js';

// Turning one resource into another - logs into boards, ore into ingots. Stock RunUO answers with a
// sound and no message either way, so the pack diff is the only evidence a conversion landed, and
// that same diff names the output's art whatever this shard numbers it.

export interface Converter {
  // Hues given up on, which is what keeps `nextStack`'s candidates shrinking and the pass count down
  writtenOff: Set<number>;

  // Returns whether the pack is clear of everything convertible. False means it stopped early - a
  // save, or the pass backstop - and the caller decides whether to come back.
  run: () => boolean;

  // Written off is not the same as impossible: a target that stepped out of range, a run of throttled
  // attempts and a stack that was briefly too small all look alike. Declines once a retry has been
  // spent with nothing converting since, which is what stops a caller that hauls the leftovers from
  // looping - `force` is for a caller that never ships the input raw and so must always reconsider.
  retry: (force?: boolean) => boolean;
}

export const createConverter = (options: {
  // Names every log line this makes, e.g. 'makeBoards' / 'smelt'
  label: string;

  // How the log lines describe what a written-off hue stays as: 'leaving it as logs'
  leftAs: string;

  attempts: number;
  timeoutMs: number;
  pollMs: number;
  delayMs: number;
  maxPasses: number;
  unskilledText: string[];

  // Silent in the pack diff and therefore indistinguishable from a material that cannot be worked, so
  // without this a run of busy moments writes off a good hue - which on a live mining run cost 86 ore
  // of a single colour.
  throttledText?: string[];

  isSaving: () => boolean;

  // Anything else that makes this pass hopeless without saying anything about the material: a fire
  // beetle that has wandered off, a forge left behind. Asked only once a stack has been chosen, so a
  // pack with nothing to convert never needs the target to exist.
  notNow?: () => string | undefined;

  // Rescanned every pass rather than planned up front: a conversion consumes the stack and creates a
  // new item, so every other serial in a snapshot goes stale the moment the first one converts.
  nextStack: (writtenOff: Set<number>) => Item | undefined;

  // Says why nothing was eligible when the pack still holds some. 'smelting freed nothing' over a
  // pack with ore in it is an accusation without evidence.
  describeSkipped?: (writtenOff: Set<number>) => string | undefined;

  // Returns false only for a cursor that never opened - everything else is judged by the pack diff.
  perform: (stack: Item) => boolean;

  // Graphics already accounted for, so the diff does not "learn" the input as an output
  known: () => Set<number>[];

  // Where a newly learned output graphic goes, and what to call it in the line that says so
  learn: (graphic: number) => void;
  learned: string;
}): Converter => {
  const writtenOff = new Set<number>();

  // Silent misses per hue, cleared by a success, so only a hue that fails repeatedly is given up on
  const misses = new Map<number, number>();

  // Whether anything has converted since the last retry. `run` refills the written-off set every
  // call, so a retry granted unconditionally answers true again on the next cycle and the caller
  // loops - which in mining is hours of 'smelting' phases before the stall watchdog ends the run.
  let progressed = true;

  // Every failing path has to come through here: one that returned without counting leaves the
  // candidate set unchanged, so the next pass picks the same stack and the loop runs to its backstop
  // instead of shrinking.
  const missed = (hue: number): void => {
    const count = (misses.get(hue) ?? 0) + 1;
    misses.set(hue, count);

    if (count >= options.attempts) {
      writtenOff.add(hue);
      log(`${options.label}: hue ${hue} failed ${count} times, ${options.leftAs}`);
    }
  };

  const learnOutput = (changes: Change[]): void => {
    for (const { key, delta } of changes) {
      if (delta <= 0) {
        continue;
      }

      const graphic = Number(key.split('/')[0]);
      if (options.known().some((set) => set.has(graphic))) {
        continue;
      }

      options.learn(graphic);
      log(`${options.label}: ${options.learned} is ${hex(graphic)}`);
    }
  };

  // The action throttle can hold a conversion well past any pause worth taking, and reading too early
  // is indistinguishable from a resource that cannot be worked - which is how ordinary logs ended up
  // on the pack animal.
  const waitForChange = (before: Counts): Change[] => {
    for (let waited = 0; waited < options.timeoutMs; waited += options.pollMs) {
      sleep(options.pollMs);

      const changes = diffCounts(before, countsByGraphic());
      if (changes.length > 0) {
        return changes;
      }
    }

    return [];
  };

  // `perform` clears the journal, so a save that starts mid-attempt is past the check at the top of
  // the pass: one cost three hues and ended a live run overweight beside a working beetle.
  const saving = (hue: number): boolean => {
    if (!options.isSaving()) {
      return false;
    }

    log(`${options.label}: the world is saving, not counting it against hue ${hue}`);

    return true;
  };

  const convertOne = (stack: Item): void => {
    const hue = stack.hue ?? 0;
    const before = countsByGraphic();

    if (!options.perform(stack)) {
      if (saving(hue)) {
        return;
      }

      // Counted like any other failure. Without this an empty hand - dead tool, just broken,
      // mid-swap - spends every pass waiting on a cursor that is never going to come.
      missed(hue);
      return;
    }

    const changes = waitForChange(before);
    if (changes.length > 0) {
      misses.delete(hue);
      progressed = true;
      learnOutput(changes);
      return;
    }

    // Asked before either wording, because a frozen shard's verdict on the material is worthless
    if (saving(hue)) {
      return;
    }

    // Nothing was attempted, so this is not a verdict on the material: left uncounted the way a world
    // save is. Checked before the unskilled wordings only because it is the commoner of the two.
    if (options.throttledText?.some((text) => journal.containsText(text))) {
      log(`${options.label}: the shard says wait, not counting it against hue ${hue}`);
      return;
    }

    // A material that needs a skill you do not have is not something retrying supplies.
    if (options.unskilledText.some((text) => journal.containsText(text))) {
      writtenOff.add(hue);
      log(`${options.label}: not skilled enough for hue ${hue}, ${options.leftAs}`);
      return;
    }

    // Otherwise it was silent, which is also what a throttled or stale attempt looks like
    missed(hue);
  };

  return {
    writtenOff,

    run: () => {
      for (let pass = 0; pass < options.maxPasses; pass++) {
        // A frozen shard answers a conversion the same way an unworkable material does - with nothing
        // at all - so without this a world save costs `attempts` and writes the hue off for the rest
        // of the run. Left for the caller: the pack is still heavy, so it comes back.
        if (options.isSaving()) {
          log(`${options.label}: the world is saving, leaving it for now`);
          return false;
        }

        const stack = options.nextStack(writtenOff);

        if (!stack) {
          const skipped = options.describeSkipped?.(writtenOff);
          if (skipped) {
            log(`${options.label}: ${skipped}`);
          }
          return true;
        }

        // Asked only once there is something that needs it: a pack with nothing to convert is
        // finished rather than blocked, which is why smeltAll can be called every time a vein dries.
        const blocked = options.notNow?.();
        if (blocked) {
          log(`${options.label}: ${blocked}, leaving it for now`);
          return false;
        }

        convertOne(stack);
        sleep(options.delayMs);
      }

      // Termination does not rest on this: a hue either converts or is given up on after `attempts`,
      // so the candidate set always shrinks.
      log(`${options.label}: hit the ${options.maxPasses} pass backstop`);
      return false;
    },

    retry: (force = false) => {
      if (writtenOff.size === 0 || (!progressed && !force)) {
        return false;
      }
      progressed = false;

      log(`${options.label}: giving ${writtenOff.size} hue(s) written off earlier another go`);
      writtenOff.clear();
      misses.clear();

      return true;
    },
  };
};

File: legion/src/uo/convert.py
import API

from uo.journal import said
from uo.pack import counts_by_graphic, diff_counts, hue_of, pack_contents


class Converter(object):
    """One resource into another, judged by the pack diff, with a per-hue write-off."""

    def __init__(self, config, log, saves):
        self._config = config
        self._log = log
        self._saves = saves
        self._written_off = set()
        self._misses = {}
        # Whether anything has converted since the last retry. Without it a retry granted
        # unconditionally answers true again next cycle and the caller loops until the watchdog ends
        self._progressed = True

    def written_off(self):
        return self._written_off

    # Every failing path comes through here: one that returns without counting leaves the candidate
    # set unchanged, so the next pass picks the same stack and the loop runs to its backstop
    def _missed(self, hue):
        count = self._misses.get(hue, 0) + 1
        self._misses[hue] = count

        if count >= self._config["attempts"]:
            self._written_off.add(hue)
            self._log("hue %d failed %d times, leaving it as %s"
                      % (hue, count, self._config["noun"]))

    # The journal is cleared to perform the conversion, so a save that starts mid-attempt is past
    # the check at the top of the pass
    def _saving(self, hue):
        if not self._saves.is_saving():
            return False

        self._log("the world is saving, not counting it against hue %d" % hue)

        if self._config["wait_on_save"]:
            self._saves.wait_out()

        return True

    # The action throttle can hold a conversion well past any pause worth taking, and reading too
    # early is indistinguishable from a resource that cannot be worked
    def _wait_for_change(self, before):
        waited = 0.0

        while waited < self._config["timeout"]:
            API.Pause(self._config["poll"])
            waited += self._config["poll"]

            gained, lost = diff_counts(before, counts_by_graphic(pack_contents()))

            if gained or lost:
                return gained

        return None

    def _convert_one(self, stack):
        hue = hue_of(stack)
        before = counts_by_graphic(pack_contents())

        if not self._config["perform"](stack):
            if self._saving(hue):
                return

            # Counted like any other failure: without this an empty hand spends every pass waiting
            # on a cursor that is never going to come
            self._missed(hue)

            return

        gained = self._wait_for_change(before)

        if gained is not None:
            self._misses.pop(hue, None)
            self._progressed = True
            self._config["learn_product"](gained)

            return

        # Asked before either wording, because a frozen shard's verdict on the material is worthless
        if self._saving(hue):
            return

        # Nothing was attempted, so this is not a verdict on the material
        if said(self._config["throttled_text"]):
            self._log("the shard says wait, not counting it against hue %d" % hue)

            return

        if said(self._config["unskilled_text"]):
            self._written_off.add(hue)
            self._log("not skilled enough for hue %d, leaving it as %s"
                      % (hue, self._config["noun"]))

            return

        # Otherwise it was silent, which is also what a throttled or stale attempt looks like
        self._missed(hue)

    def run(self):
        for _pass in range(self._config["passes"]):
            # A frozen shard answers a conversion the same way an unworkable material does, so
            # without this a world save costs the attempts and writes the hue off for the run
            if self._saves.is_saving():
                self._log(self._config["saving_message"])

                return False

            stack = self._config["next_source"](self._written_off)

            if stack is None:
                self._config["nothing_to_do"](self._written_off)

                return True

            # Asked only once there is something that needs it, which is what makes converting on
            # every dry spot affordable
            blocked = self._config["blocked"]()

            if blocked is not None:
                self._log("%s, leaving it for now" % blocked)

                return False

            self._config["about_to_convert"]()
            self._convert_one(stack)
            API.Pause(self._config["delay"])

        self._log("hit the %d conversion pass backstop" % self._config["passes"])

        return False

    def retry_written_off(self, force=False):
        if len(self._written_off) == 0 or (not self._progressed and not force):
            return False

        self._progressed = False
        self._log("giving %d hue(s) written off earlier another go" % len(self._written_off))
        self._written_off.clear()
        self._misses.clear()

        return True
